specs/Alert_Notifications.adoc

```diff
@@ -0,0 +1,189 @@
+
+
+= Alert Notifications
+
+Alerting system is essential to spot the problem and notify it to the corresponding user.
+The thresholds for alerts will be configured in grafana and the alerts will be raised by grafana.
+These alerts are captured by monitoring-integration module and notified with proper 
+details to respective user via the notifier.
+
+
+== Problem description
+
+The present alerting system monitors all the events and retrieves alert specific events.
+The alert specific events are processed and stored again in etcd under /alerting/alerts/
+as alerts recognized by tendrl. This leads to excess read and writes to etcd.
+Present implementation causes the alerting component to process events even
+when it can be done by a separate component before inserting the event in etcd.
+
+== Use Cases
+
+Goal: Notify user in case of any alert event.
+
+Actors:
+
+1) User
+
+Steps:
+
+* Node Agent stores alerting events in etcd.
+
+* These events will be fetched by the Notifier.
+
+* The alerting event is to be sent to respective user via configured
+  notification channels.
+
+== Proposed change
+
+
+Present Alert system(https://github.com/Tendrl/alerting) will be renamed as “Notifier”
+
+The main functionality of Notifier would be to:
+
+Monitor etcd only for alert specific events.
+
+Sending alert notifications to configurable receiving end-points.
+
+Presently the alerting module processes all events under messages\events
+in etcd and looks for events having priority as “notice” which signifies they are
+alerts. Instead of processing all events an AlertManager will be configured to
+receive just alerting events and notify users about these events.
+
+For this the notifier must receive only the alerting events.
+This can be achieved in 2 ways:
+
+* All alerting events will be stored in etcd under alerting/alerts/{cluster,node} by
+  the node-agent as per the scope of the event (cluster level scope or node level scope).
+  A watcher(AlertWatcher) will be configured to poll etcd (alerting/alerts) for any new
+  events. On detecting new events the evetns will be passed to notifier which would
+  retrieve payload from the event, parse it and using the configured notification channel
+  would notify the users of these events.
+
+ Advantage:
+ 
+     ** Simple architecture and easy to implement.
+
+  Drawback:
+ 
+      ** Polling etcd for new alerts is not efficient.
+ 
+* To avoid polling etcd, node agent instead of just logging the events to
```

Comment:
The logic behind setting the alert conditions first is that it is vital to ensure that the system knows about a condition before notifying it to the end user. Routing this via the notifier creates an extra hop in this loop and introduces a higher chance of the condition being lost.

In addition, the notifier's job is to only notify. It must not have write access to etcd in an important global directory. That's why it is the node agent's responsibility, as the transport for all the alert conditions, to set the alert condition first and foremost.

For this particular logic to work, the node agent would have to both set the alert condition and hand the said condition to the notifier.